ui/src/app/services/error-reporter.service.ts

```diff
@@ -31,26 +31,33 @@ export class ErrorReporterService extends ErrorHandler {
     });
   }
 
-  handleError(error: any) {
-    // Always log to console regardless of whether Stackdriver is enabled.
-    super.handleError(error);
-    if (!this.stackdriverReporter) {
+  /**
+   * Unwrap to avoid vague top-level errors like: "Promise rejected: {}". This method
+   * mutates the incoming error in order to allow preservation of the original stack trace.
+   */
+  private async expandErrorMessage(error: any): Promise<any> {
+    if (!error.rejection || !(error.rejection instanceof Response)) {
       return;
     }
 
-
-    // We want to avoid sending XHR errors to stackdriver because they should
-    // already be being logged by the API with the stacktrace and information
-    // we care about.
-
-    // xhrError is set to true by the interceptor. The rejection piece handles
-    // return values from promises that error out.
-    if (error.rejection && error.rejection.xhrError) {
+    const httpResp = error.rejection;
+    const prefix = `${httpResp.status} @ ${httpResp.url}`;
+    if (httpResp.bodyUsed) {
+      error.message = prefix;
       return;
     }
-    // The check on xhrError explicitly handles error values that are returned from direct
-    // api errors, rather than promises.
-    if (error.xhrError) {
+    const json = JSON.stringify(await httpResp.json());
+    error.message = `${prefix}: ${json}`;
+    return;
+  }
+
+  async handleError(error: any) {
```

Comment:
Does changing this to `async` have any impacts that would be relevant - like changing the order that the error messages appear in, in the log?

Reply:
Good question. I don't think it's a problem here, my reasoning follows:

- This particular methods gets triggered when an error hits the top of the stack (or at least, top of the Angular harness). So at this point, we're outside of our normal code flow.
- Because this also gets hit when an error is thrown, it's also not likely that we're depending on two errors coming in from one logical piece of code - to the point where we'd care about the order

Practically speaking, I don't think the order typically matters, and this is purely for manual debugging anyways. The response body on an error also should not be streaming in or anything, from any typical service. I expect in all cases that promise will resolve immediately (though this still affects the flow control).